website/util/metrics.py

```diff
@@ -1,14 +1,119 @@
 # -*- coding: utf-8 -*-
-
+from enum import Enum
 
 def get_entry_point(user):
     """
     Given the user system_tags, return the user entry point (osf, osf4m, prereg, institution)
     In case of multiple entry_points existing in the system_tags, return only the first one.
     """
-    entry_points = ['osf4m', 'prereg_challenge_campaign', 'institution_campaign', 'osf_preprints']
+    entry_points = [
+        CampaignSourceTags.Osf4m.value,
+        CampaignSourceTags.PreregChallenge.value,
+        'institution_campaign',
+        ProviderSourceTags.OsfPreprints.value
+    ]
     for i in user.system_tags:
         if i in entry_points:
             return i
     else:
         return 'osf'
+
+
+def provider_source_tag(provider_id, service=None):
+    if service:
+        return 'source:provider|{}|{}'.format(service, provider_id)
+    else:
+        return 'source:provider|{}'.format(provider_id)
+
+
+def campaign_source_tag(campaign_name):
+    return 'source:campaign|{}'.format(campaign_name)
+
+
+def provider_claimed_tag(provider_id, service=None):
+    if service:
+        return 'claimed:provider|{}|{}'.format(service, provider_id)
+    else:
+        return 'claimed:provider|{}'.format(provider_id)
+
+
+def campaign_claimed_tag(campaign_name):
+    return 'claimed:campaign|{}'.format(campaign_name)
+
+
+# Needs to be updated when new providers are added.
```

Comment:
Is it possible to do this implicitly everywhere, as in `Preprint.add_unregistered_contributor`? We generally try to not block Product when adding new providers. All the info is there already -- `provider._id`, `provider.readable_type` (though that would lead to `registration` rather than `registry`).

If not, please ensure that product knows and their provider onboarding docs are updated before this goes out.

Reply:
What I meant here is that if there are new providers added to OSF prior to the normalization script is ran, we should add them to this list. After that, all new providers added will automatically have there source tags created implicitly when a unregistered contrib is added to one of its preprints.

Reply:
Yeah, I realized last night that `indiarxiv_preprints` is missing.

But if it's the case that non-OSF provider-specific tags are only to be used for the data migration, why are they defined here, rather than just in one of the two scripts? It seems like it will lead to maintenance confusion later.

Reply:
Umm. I thought separating the non-OSF specific tags from the rest creates more confusion. And putting all the source tag definitions in one placed looked like a nice idea.
Maybe I can update the code comment to say something like "non-OSF tags are used for the normalization script, and does not need to be updated to include new providers when they are added to OSF"?

Reply:
If the only reason to define them is for clarity in the scripts, I'm not seeing why they need to be defined anywhere other than the scripts. I might be missing something. It's just a list of things to keep updated with new _database objects_, even though running application code doesn't need it, no?